nodejs/week2/homework/meal-sharing/src/backend/app.js

```diff
@@ -0,0 +1,102 @@
+const express = require("express");
+const app = express();
+
+const mealsRouter = require("./api/meals-router");
+app.use("/api/meals", mealsRouter);
+
+const meals = require("./data/meals.json");
+const reservations = require("./data/reservations.json");
+
+const reservationsRouter = require("./api/reservations-router");
+app.use("/api/reservations", reservationsRouter);
+
+const reviewsRouter = require("./api/reviews-router");
+const reviews = require("./data/reviews.json");
+app.use("/api/reviews", reviewsRouter);
+
+// app.use binds middleware to your application. You can give app.use a path and router. The mini router will take care of all requests with the path
+mealsRouter.get("/:id", (req, res) => {
+  const id = req.params.id;
+  const filteredMeal = meals.filter((meal) => meal.id === parseInt(id));
+  res.send(filteredMeal);
+});
+
+//Respond with the json for all the meals
+// mealsRouter.get("/", (req, res) => {
+//   res.send(meals);
+// });
+
+//Respond with the json for the reservation with the corresponding id *
+reservationsRouter.get("/:id", (req, res) => {
+  const filteredReservation = reservations.filter(
+    (reservation) => reservation.id === parseInt(req.params.id)
+  );
+  res.send(filteredReservation);
+});
+
+//Respond with the json for all reservations *
+reservationsRouter.get("/", (req, res) => {
+  res.send(reservations);
+});
+
+// Respond with the json for the review with the corresponding id *
+reviewsRouter.get("/:id", (req, res) => {
+  const filteredreview = reviews.filter(
+    (review) => review.id === parseInt(req.params.id)
+  );
+  res.send(filteredreview);
+});
+
+// Respond with the json for all reviews *
+reviewsRouter.get("/", (req, res) => {
+  res.send(reviews);
+});
+let limitedMeals = [];
+//Get meals that has a price smaller than maxPrice
+mealsRouter.get("/", (req, res) => {
+  let data = meals;
+  console.log("String for testing");
+  console.log(req.query);
+
+  if ("maxPrice" in req.query) {
```

Comment:
This code becomes hard to understand due to the usage of `if`, combined with filters and a shared state that is overwritten in each `if` block.

Try to re-write this using chaining instead.
```js
   function createMaxPriceFilter(rawMaxPrice) {
     // Do input validation
     const maxPrice = parseInt(rawMaxPrice);
     const priceLimit = isNaN(maxPrice) ? Number.MAX_VALUE : maxPrice;
     
     // Return a function that can do the filter
     return (meal) => meal.price <= priceLimit;
   }

  const matchingMeals = meals
    .filter(createMaxPriceFilter(req.query.maxPrice))
    .filter(createTitleFilter(req.query.title))
```

Another approach is to first validate all the input parameters in a separate function and then use the validated values in the filter functions.